archunit/src/main/java/com/tngtech/archunit/library/GeneralCodingRules.java

```diff
@@ -449,29 +443,34 @@ public static ArchRule testClassesShouldResideInTheSamePackageAsImplementation(S
 
     private static ArchCondition<JavaClass> resideInTheSamePackageAsTheirTestClasses(String testClassSuffix) {
         return new ArchCondition<JavaClass>("reside in the same package as their test classes") {
-            Map<String, JavaClass> testClassesBySimpleClassName = new HashMap<>();
+            Map<String, List<JavaClass>> testClassesBySimpleClassName = new HashMap<>();
 
             @Override
             public void init(Collection<JavaClass> allClasses) {
                 testClassesBySimpleClassName = allClasses.stream()
                         .filter(clazz -> clazz.getName().endsWith(testClassSuffix))
-                        .collect(toMap(JavaClass::getSimpleName, identity()));
+                        .collect(groupingBy(JavaClass::getSimpleName));
             }
 
             @Override
             public void check(JavaClass implementationClass, ConditionEvents events) {
                 String implementationClassName = implementationClass.getSimpleName();
                 String implementationClassPackageName = implementationClass.getPackageName();
                 String possibleTestClassName = implementationClassName + testClassSuffix;
-                JavaClass possibleTestClass = testClassesBySimpleClassName.get(possibleTestClassName);
+                List<JavaClass> possibleTestClasses = testClassesBySimpleClassName.getOrDefault(possibleTestClassName, emptyList());
+
+                if (possibleTestClasses.isEmpty()) {
```

Comment:
Why not keep this in the declaration of `isTestClassInWrongPackage` and avoid the early return?